### PR TITLE
issue #97 - remove dependence on Bouncy Castle

### PR DESCRIPTION
**Pull request by travisspencer**
This PR removes the dependency on Bouncy Castle by using Java SE functionality and existing dependencies (namely, Guava) as a replacement for the things that it was doing before.

### CODE REVIEW

File: webauthn-server-core/src/test/scala/com/yubico/webauthn/RelyingPartyAssertionSpec.scala [hunk 1]
```diff
@@ -66,14 +69,16 @@ import scala.util.Success
...
+class RelyingPartyAssertionSpec extends FunSpec with Matchers with ScalaCheckDrivenPropertyChecks with BeforeAndAfterAll {
 
   private def jsonFactory: JsonNodeFactory = JsonNodeFactory.instance
-  private val crypto = new BouncyCastleCrypto()
+  private val crypto = new Crypto()
 
   private def sha256(bytes: ByteArray): ByteArray = crypto.hash(bytes)
   private def sha256(data: String): ByteArray = sha256(new ByteArray(data.getBytes(Charset.forName("UTF-8"))))
 
+  override def beforeAll() = Security.addProvider(new BouncyCastleProvider())
```
Comment (travisspencer) [inline webauthn-server-core/src/test/scala/com/yubico/webauthn/RelyingPartyAssertionSpec.scala:80]: Those who want to use Bouncy Castle should do this or other equivalent (e.g., by editing `$JAVA_HOME/lib/security/java.security` or passing command line flags).

**Comment by travisspencer** [comment]
All these tests are passing with the JRE we use (Azul Zulu 8 Enterprise with OpenJSSE and OpenDSSA) without loading Bouncy Castle.

**Comment by emlun** [comment]
Hi! This definitely seems worthwhile to do, but unfortunately I don't have the time to look into this right now. Sorry, I'll let you know when I do.

**Comment by travisspencer** [comment]
All good, @emlun. FYI though, we hope to ship this in our next, next release which will be in 9 weeks. I hope time frame works.

**Comment by travisspencer** [comment]
It would be very helpful for us if a release with this change was published within a week or so. Would that be possible?

**Comment by emlun** [comment]
Sorry for the late response, and sorry for missing your deadline. Now I am finally able to focus on this full time, so we should be able to finish this up within two weeks at the very latest. I hope this hasn't caused too much trouble for you.

At a general level this looks pretty good. Some of the tests fail in JDK 11 with `NO_BOUNCY_CASTLE` due to EdDSA not being supported, but it looks like JDK 15 does include support for that. Still, we might have to make this a major version bump since it breaks some backwards compatibility.

In JDK the tests instead fail because `secp256k1` is no longer supported - but that is in a test to verify that `secp256k1` keys are _not_ accepted as U2F attestation certificates, so that should also be easy to fix.

Thanks for working on this, and sorry for the delay. We should be able to get this released pretty soon.

**Comment by travisspencer** [comment]
We're managing and understand, @emlun , that you're busy with a lot of things.

About Java 11: wouldn't people using that continue to use Bouncy Castle? If so, maybe the tests need a touch up only? When I look at the [Java 11 build report](https://github.com/Yubico/java-webauthn-server/pull/98/checks?check_run_id=1669455201), I don't see the failures, so I'm not entirely sure what you're referring to. If you can point that out in more detail, I might be able to help a bit.

Happy to do more; just holler if you need anything.

**Comment by emlun** [comment]
>About Java 11: wouldn't people using that continue to use Bouncy Castle?

Yes, that seems like the best option. I did some experiments with [ed25519-java](https://github.com/str4d/ed25519-java) as well, but there's a bit of mismatch there because that provider uses different algorithm names than BouncyCastle and [OpenJDK 15 do](https://bugs.openjdk.java.net/browse/JDK-8190219).

It currently looks like the most likely path here is to keep BouncyCastle as a dependency throughout the 1.x major version - since dropping it would require manual intervention from JDK 11 users to maintain compatibility - but make the library continue anyway if BouncyCastle cannot be loaded, and instead fail when for example `Signature.getInstance("EdDSA")` gets called. This way it should be possible for users to exclude BouncyCastle from the downstream classpath and use a different provider, while we maintain function for users that don't actively opt out in this way. Then we'll look at dropping the dependency for a would-be future 2.0 release. Would that work for you?

>When I look at the [Java 11 build report](https://github.com/Yubico/java-webauthn-server/pull/98/checks?check_run_id=1669455201), I don't see the failures, so I'm not entirely sure what you're referring to.

Yeah, the failure only happens when run with `USE_BOUNCY_CASTLE=false` (previously `NO_BOUNCY_CASTLE=true`), which the GitHub Actions workflows didn't do at the time. Sorry for the noise, I've been trying to get the same failure to appear in Actions but having some issues with various version incompatibilities with OpenJDK, Gradle and Lombok. But anyway, you should be able to see the failure if you run the tests locally in OpenJDK 11 with `USE_BOUNCY_CASTLE=false ./gradlew clean test`.

And sorry, I see now that my previous comment should have read "In JDK **15** the tests instead fail because `secp256k1` is no longer supported". And to be clear, that is a defect in the tests, not the library. It should be easy to fix, I'm on it.

>Happy to do more; just holler if you need anything.

Thanks! No need right now apart from the above question about release strategy.

**Comment by travisspencer** [comment]
> that provider uses different algorithm names than BouncyCastle and [OpenJDK 15 do](https://bugs.openjdk.java.net/browse/JDK-8190219).

Could the name be abstracted out? This was the case with RSA-PSS as well, and it was possible to handle by abstraction of the name in that case.

> This way it should be possible for users to exclude BouncyCastle from the downstream classpath and use a different provider, while we maintain function for users that don't actively opt out in this way.

This does make me nervous. We can put test in place to ensure that BouncyCastle isn't accidentally brought in, but I prefer an opt-in than opt-out mechanism. This crypto stuff brings issues with import/export/re-export, patents, etc. so it's not something to get wrong. That's why I'd prefer an opt-in. I get that it's a change though, but it would be great if we can find a way to drop the dependency entirely and allow those that want/need it to add it explicitly. They're presumable OK with the associated risks if they explicitly bring in this dependency.

**Comment by emlun** [comment]
>Could the name be abstracted out? This was the case with RSA-PSS as well, and it was possible to handle by abstraction of the name in that case.

I think the first step here is rather to open an issue in ed25519-java to consider changing the name; I'll do that. I'm guessing it's in their interest to align with the standard JCA name anyway. Either way BouncyCastle still works for JDK 11, so for java-webauthn-server we'll stick with that for now - that also spares us from having to evaluate another dependency to add or recommend.

>This does make me nervous. We can put test in place to ensure that BouncyCastle isn't accidentally brought in, but I prefer an opt-in than opt-out mechanism. [...]

Good points. We could split out a new "core-minimal" module (placeholder name) without the BouncyCastle dependency, and make the current "core" a meta-module that just bundles "core-minimal" with BouncyCastle. Then you could depend on "core-minimal" instead, and for 2.0 we'd make that the new "core". Would that work for you?

We'll have to think a bit more about the details, for example what it means for transitive dependencies like that from `webauthn-server-attestation`, but that seems to me like a promising solution.

**Comment by travisspencer** [comment]
> We could split out a new "core-minimal" module (placeholder name) without the BouncyCastle dependency

This sounds like a great solution. Suggested names: 

* inner-core 
* nucleolus (the core of a core)
* mantle 

Of these, I like inner-core the best.

**Comment by emlun** [comment]
Hm. I started writing a reply to your last comment, but apparently I never posted it. Thanks for the naming suggestions! - but we went with `core-minimal` in the end. I hope you didn't mind me hijacking your branch...

Anyway, after some more internal design discussion we've arrived at the solution in commit 35ce19c: as described above, we've renamed the `webauthn-server-core` module to `webauthn-server-core-minimal` and removed the BouncyCastle (BC) dependency, and added a new `webauthn-server-core` dependency that has `webauthn-server-core` and BC as its dependencies. Internally, `webauthn-server-core-minimal` still has a compile-time dependency on BC and will [attempt to fall back to the BC provider when necessary](https://github.com/Yubico/java-webauthn-server/blob/35ce19cd395681db4751a8b55c8d9871327ca2f1/webauthn-server-core/src/main/java/com/yubico/webauthn/Crypto.java#L91-L102) - but BC is not included in the runtime dependencies, and the library will not attempt to load it if the existing JCA providers suffice.

The `webauthn-server-attestation` module still has a direct dependency on BC, but for the ASN.1 parser rather than the JCA provider. Please let us know if you need that decoupled too.

To ensure that the packaging is correct, there are new tests in [test-dependent-projects/java-dep-webauthn-server-core-minimal](https://github.com/Yubico/java-webauthn-server/tree/35ce19cd395681db4751a8b55c8d9871327ca2f1/test-dependent-projects/java-dep-webauthn-server-core-minimal) that ensure both that BC is not loaded as a JCA provider, and that it's not present in the classpath for downstream projects that depend only on the `core-minimal` module.

You are welcome to verify that these tests fail if you add BC as a runtime dependency of `core-minimal` (note that the `core-minimal` module is still hosted in the `webauthn-server-core/` directory, to avoid some file renames we would later revert anyway), and that `./gradlew generatePomFileForJarsPublication` generates POMs with the expected dependency declarations.

Please try this out and see if it works for you! Let us know if you'd like us to publish a proper release candidate build, and we can do that. Otherwise, our current plan is to publish this as release 1.8.0 on 2021-03-15. Beyond that the plan is to fully drop the BC dependency, and again make `core-minimal` the main `core` module, in a future 2.0 release.

**Comment by travisspencer** [comment]
>  I hope you didn't mind me hijacking your branch...

np 

> The `webauthn-server-attestation` module still has a direct dependency on BC, but for the ASN.1 parser rather than the JCA provider. Please let us know if you need that decoupled too.

We can't have BC _at all for anything_. The patent, import/export issues, indemnities, indemnity insurance, escrow, OSS attribution, etc., etc. go up with _every_ dependency. BC is one that comes with a big price in all aspects, so it's not an option for us for any purpose. Besides, if the the ASN.1 parsing your talking about is for that GUID, that was super simple stuff that doesn't require a lib IMHO. Can it please be removed entirely? (Maybe you're talking about something else though, I'm not as familiar with all these modules as you are.) 

> To ensure that the packaging is correct, there are new tests in test-dependent-projects/java-dep-webauthn-server-core-minimal that ensure both that BC is not loaded as a JCA provider, and that it's not present in the classpath for downstream projects that depend only on the core-minimal module.

Can this be run on GitHub using [Azul's Zulu community edition](https://www.azul.com/downloads/zulu-community/?package=jdk) and their [OpenDSA library](https://github.com/openjsse/openeddsa)? This will give us 100% assurance that our rig is good to go. Happy to add that if you can give me a pointer or two on how to.

**Comment by dainnilsson** [comment]
> > The `webauthn-server-attestation` module still has a direct dependency on BC, but for the ASN.1 parser rather than the JCA provider. Please let us know if you need that decoupled too.
> 
> We can't have BC _at all for anything_ (that we ship; use in tests is OK). The patent, import/export issues, indemnities, indemnity insurance, escrow, OSS attribution, etc., etc. go up with _every_ dependency. 

I think the question is more on whether or not you intend to use the `webauthn-server-attestation` module or not, as it's optional. If you do, then we need to decouple there as well. If not, then at least there's no urgency there.

**Comment by travisspencer** [comment]
> I think the question is more on whether or not you intend to use the `webauthn-server-attestation` module or not

We haven't used it yet, and I'm not sure what it is, honestly. Is this something to verify an attestation created from an Android/iOS device independent of WebAuthn? If any event, we may use this down the line but we haven't looked at it for what we're planning in our coming release. So, I think we're good.

**Comment by emlun** [comment]
Yeah, if you don't have an explicit dependency on `com.yubico:webauthn-server-attestation` you should be good. That module is where we were going to put integrations with attestation metadata sources such as the FIDO MDS, but we haven't done that yet. So currently it's mostly a placeholder that comes preloaded with Yubico attestation root certificates.

So to be abundantly clear here, the current implementation state is:

- `webauthn-server-core-minimal` will still have a _compile-time only_ dependency on BC. There will be explicit references to BC code in the released artifact.
- `webauthn-server-core-minimal` will not have a _runtime_ dependency on BC. So no BC code will be present in your runtime classpath if you depend only on `webauthn-server-core-minimal`.

>Can this be run on GitHub using [Azul's Zulu community edition](https://www.azul.com/downloads/zulu-community/?package=jdk) and their [OpenEdDSA library](https://github.com/openjsse/openeddsa)?

I'll look into it!

**Comment by travisspencer** [comment]
Sounds perfect @emlun . Thanks for the clear and detailed explanation. 

>> Can this be run on GitHub using Azul's Zulu community edition and their OpenEdDSA library?
>
> I'll look into it!

Will run some tests of my own and report back, but would *greatly* appreciate it if the above can also be set to run here.

**Comment by emlun** [comment]
Looks like the `actions/setup-java` action actually [uses the Zulu build by default](https://github.com/actions/setup-java#setup-java), so it seems like we're already covered there. The tests [pass on JDK 8 with OpenEdDSA](https://github.com/Yubico/java-webauthn-server/actions/runs/647538972), but it seems like OpenEdDSA is currently [incompatible with JDK versions other than JDK 8](https://github.com/openjsse/openeddsa/issues/2). There's also a [minor issue with the packaging](https://github.com/openjsse/openeddsa/issues/1).

So with these issues in mind I don't think we'll be running the GitHub tests with OpenEdDSA before the 1.8 release on Monday, but we'll definitely revisit it if at least the packaging issue is fixed.

**Comment by travisspencer** [comment]
The Azul gang has supplied us an OpenEdDSA JAR to use with v. 8. Let me see if I can get some help from them on this.

**Comment by emlun** [comment]
Thanks, but I don't want to distribute the JAR within this repository and I don't want to add special-case handling just for this dependency. I did push [a commit with the embedded JAR file](https://github.com/Yubico/java-webauthn-server/commit/f2c88ab2281287cd9f5bb582be5904baad309f54) to confirm that [the tests pass with OpenEdDSA](https://github.com/Yubico/java-webauthn-server/actions/runs/647538972), but that commit will remain unreachable from the development branches. We'll resume testing with OpenEdDSA once the packaging is fixed, but I'll put that on hold for now.

**Comment by emlun** [comment]
But with that, we are ready to release this. Thanks for your help with both implementation and design! And again sorry for taking so long on this. Expect a 1.8.0 release later today, and please let us know if you have any problems!

**Comment by travisspencer** [comment]
Thanks, @emlun, for the help and the new release. We can circle back on the OpenEdDSA stuff later. Great work on the release and much appreciation.

**Pull request merged**